src/action/mod.rs

```diff
@@ -35,90 +35,176 @@ impl Finish {
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub(crate) enum Patch {
+    /// Replace the area spanned by `replace` with `replacement`.
+    /// Since `Span` is inclusive, `Replace` always will replace a character in the original sources.
+    Replace {
+        replace_span: Span,
+        replacement: String,
+    },
+    /// Location where to insert.
+    Insert {
+        insert_at: LineColumn,
+        content: String,
+    },
+}
 
-        if let Some(ref bandaid) = nxt {
-            if !bandaid.span.covers_line(line_number) {
-                sink.write(content.as_bytes())?;
-                sink.write("\n".as_bytes())?;
-                continue;
-            }
+impl<'a> From<&'a BandAid> for Patch {
+    fn from(bandaid: &'a BandAid) -> Self {
+        // TODO XXX
+        Self::from(bandaid.clone())
+    }
+}
+
+impl From<BandAid> for Patch {
+    fn from(bandaid: BandAid) -> Self {
+        match bandaid {
+            bandaid if bandaid.span.start == bandaid.span.end => Self::Insert {
+                insert_at: bandaid.span.start,
+                content: bandaid.replacement,
+            },
+            _ => Self::Replace {
+                replace_span: bandaid.span,
+                replacement: bandaid.replacement,
+            },
         }
+    }
+}
 
-        let content_len = content.chars().count();
-        while let Some(bandaid) = nxt.take() {
-            trace!("Applying next bandaid {:?}", bandaid);
-            trace!("where line {} is: >{}<", line_number, content);
-            let range: Range = bandaid
-                .span
-                .try_into()
-                .expect("There should be no multiline strings as of today");
-            // write prelude for this line between start or previous replacement
-            if range.start > remainder_column {
-                sink.write(util::sub_chars(&content, remainder_column..range.start).as_bytes())?;
-            }
-            // write the replacement chunk
-            sink.write(bandaid.replacement.as_bytes())?;
-
-            remainder_column = range.end;
-            nxt = bandaids.next();
-            let complete_current_line = if let Some(ref bandaid) = nxt {
-                // if `nxt` is also targeting the current line, don't complete the line
-                !bandaid.span.covers_line(line_number)
+/// Correct all lines by applying bandaids.
+///
+/// Assumes all `BandAids` do not overlap when replacing.
+/// Inserting multiple times at a particular `LineColumn` is ok,
+/// but replacing overlapping `Span`s of the original source is not.
+///
+/// This function is not concerend with _any_ semantics or comments or
+/// whatsoever at all.
+fn correct_lines<'s, II, I>(patches: II, source_buffer: String, mut sink: impl Write) -> Result<()>
+where
+    II: IntoIterator<IntoIter = I, Item = Patch>,
+    I: Iterator<Item = Patch>,
+{
+    let patches = patches.into_iter();
+    let mut patches = patches.peekable();
```

Comment:
Nitpick: what's wrong with putting it together like this?

```suggestion
    let mut patches = patches.into_iter().peekable();
```